### PR TITLE
ssl model refactoring remove models, provide full examples instead

### LINKED ISSUE #483: SSL model refactoring: remove models, provide full examples instead
## Current state:

Currently, we offer 3 levels of SSL parts:
- low-level: the separate building blocks such as a projection head, nn memory bank, backbone creation
https://github.com/lightly-ai/lightly/blob/8dde4d68c0598758bfd05c05f225bf754d9f1ed5/lightly/models/modules/heads.py#L12
- mid-level: the model (backbone + heads + forward functions)
https://github.com/lightly-ai/lightly/blob/8dde4d68c0598758bfd05c05f225bf754d9f1ed5/lightly/models/moco.py#L13
- high-level: the model, criterion, training loop put together as pytorch lightning module
https://github.com/lightly-ai/lightly/blob/8dde4d68c0598758bfd05c05f225bf754d9f1ed5/docs/source/getting_started/benchmarks/cifar10_benchmark.py#L144

## Target state:
We remove the mid-level representations (delete that code). We provide more examples on the high-level: One example pytorch lightning module for every implemented paper

## Advantages:
- As we don't offer mid-level blocks, we don't need to document and maintain them and try to keep them flexible.
- We offer full-fletched examples of how to put the building blocks together to a complete pytorch lightning module.
- By calling the high-level module an example, we don't need to make it configurable and can keep using hardcoded parameters. 
- The high-level example shows directly (not nested) all building blocks and how they play together, thus our users can easily adapt the example on all levels to their needs. They can also change some building blocks very easily and thus come up with new architectures.

### PR DESCRIPTION
**Pull request by MalteEbner**
closes #483

## Description:
- implemented high-level examples for MoCo and SimCLR as pytorch lightning modules
- put the MomentumEncoder is an independent class (no Mixin anymore)

## Open to-dos:

- [ ] write new momentum encoder (make it more flexible)
- [ ] make cli work 
- [ ] deprecation warning old momentum encoder
- [ ] implement high-level example for one new model (e.g. SwAV) https://github.com/lightly-ai/lightly/issues/492
- [ ] deprecation warnings for old models

- [ ] 3. delete mid-level models in lightly/lightly/models
- [ ] 4. ensure compatibility with old pretrained models
- [ ] 5. update documentation
- [ ] 6. implement high-level examples for all implemented papers based on base building blocks

### CODE REVIEW

File: lightly/models/examples/moco_system.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
+import pytorch_lightning as pl
+import torch
+
+from lightly.models.modules.heads import MoCoProjectionHead
+
+import lightly
+from lightly.models.modules.momentum_encoder import MomentumEncoder
+
+
+class MocoSystem(pl.LightningModule):
```
Comment (MalteEbner) [inline lightly/models/examples/moco_system.py:10]: I would keep calling it MocoSystem for the moment as the name MoCo is already taken by https://github.com/lightly-ai/lightly/blob/8dde4d68c0598758bfd05c05f225bf754d9f1ed5/docs/source/getting_started/benchmarks/cifar10_benchmark.py#L144

**Comment by IgorSusmelj** [comment]
That does look good to me. I'd need to play around a bit to be able to give more feedback.

File: lightly/models/examples/moco_system.py (hunk 1 above)
Comment (philippmwirth) [reply lightly/models/examples/moco_system.py:10]: I am also in favor of **not** calling it `MoCoSystem`.

File: lightly/models/examples/moco_system.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    def __init__(self, dataloader_kNN, num_classes):
+        super().__init__(dataloader_kNN, num_classes)
+        # create a ResNet backbone and remove the classification head
+        resnet = lightly.models.ResNetGenerator('resnet-18', num_splits=8)
```
Comment (philippmwirth) [inline lightly/models/examples/moco_system.py:14]: I hate that the parameters are hard-coded here and below. IMO this is not an option except if we move the `examples` directory to the top-level (same level as `docs` and `tests`).

Alternatives:
- Pass them as arguments
- Pass `config` objects as arguments to avoid too many arguments
    > "If you have a procedure with 10 parameters, you probably missed some." ~ [Perlisims](http://www.cs.yale.edu/homes/perlis-alan/quotes.html)
- Pass projection head, backbone, loss, and optimizer as arguments to the constructor (they have their own default parameters which we could use)
- We brainstorm and try to find another solution

File: lightly/models/examples/moco_system.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+import torch
+
+from lightly.models.modules.heads import MoCoProjectionHead
+
+import lightly
+from lightly.models.modules.momentum_encoder import MomentumEncoder
+
+
+class MocoSystem(pl.LightningModule):
+    def __init__(self, dataloader_kNN, num_classes):
```
Comment (philippmwirth) [inline lightly/models/examples/moco_system.py:11]: The `dataloader_kNN` and `num_classes` are just a copy-paste artefact, right?

File: lightly/models/examples/moco_system.py (hunk 2 above)
Comment (MalteEbner) [reply lightly/models/examples/moco_system.py:14]: If we have some arguments or even a config object, we have the following disadvantages:
- more maintenance and documentation work
- It is less transparent which parameters are used where and why. E.g. will there be one config object for all models or one for each model? Can I pass a config object with `momentum=0.999` also to the SimCLR Model?
- We can hardly cover all parameters. E.g. what about the scheduler and its parameters? What about using or not using a symmetric loss?
- We become similar to pytorch lightning bolt. 

I agree that we can easily put the examples to the top-level or as an own subpart of the documentation. I like the idea of brainstorming a little more on that.

File: lightly/models/examples/moco_system.py (hunk 3 above)
Comment (MalteEbner) [reply lightly/models/examples/moco_system.py:11]: Yes, sorry

File: lightly/models/examples/moco_system.py (hunk 2 above)
Comment (MalteEbner) [reply lightly/models/examples/moco_system.py:14]: One general problem might be that we don't have a clear target group of the pip package yet. This is also covered in another issue: https://github.com/lightly-ai/lightly/issues/479

Perhaps we should have a discussion about who we want to target with our pip package.

**Comment by MalteEbner** [comment]
Closed, we now know more clearly what to do with which steps: https://github.com/lightly-ai/lightly/issues/493
